content/posts/2019-11-07-ln-split-tickets-decred-blog.md

```diff
@@ -0,0 +1,159 @@
+
+This tree of transactions can be prepared in advance during the setup stage of the multi-owner ticket and participants can hold onto it and broadcast the relevant branches in case of non-cooperation.
+
+However the most important open question was: how can we allow partial redeeming of this prebuilt transaction off-chain, such that the tree (or a large portion of it) doesn't need to be published on-chain?
+
+We have now come up with a design to solve this and will provide it in the rest of this post.
+
+# MRTTREE 
+
+Start by observing that an HTLC preimage is simply a 32 bit random number. And secp256k1 private keys, the ones used by Decred and Bitcoin to sign transactions are also for the most part just 32 bit random numbers.
+
+We can combine these two uses of a single number in a clever way by constructing the tree of transactions such that the leaf outputs are redeemable by a standard HTLC (that is to say, by presenting the preimage `x` to `Hash(x)`) and then ensuring that this exact same `x` is the key that is used to sign the previous corresponding branches of the tree.
+
+This allows off-chain redeeming of funds by using a collaborating third party called a _provider_: the provider will keep the on-chain funds while paying back atomically off-chain. Assuming each individual user is the only one in possession of the preimage that encumbers each leaf output, once they redeem their funds off-chain by presenting the preimage to the provider then the provider is able to rewrite the branches of the tree of transactions such that the funds are sent to it.
+
+{{< legendfigure src="/images/posts/ln-splits-02-mrttree-partial02.png" alt="MRTTREE with partial redemption" title="Partial off-chain redemption of the tree of transactions by users and on-chain by provider." >}}
+
+In the limit where all participants collaborate, only a single transaction is sent on-chain and all funds are dispersed off-chain. With partial collaboration only the non-rewritten branches are sent on-chain minimizing the load on the blockchain and the effective fees paid. And given that each user can claim their funds individually, there's no need for real time collaboration between all users at the exact same time - until the root transaction is published there's ample time for any individual user to redeem their funds off-chain and therefore not pay the higher on-chain fees.
+
+In order to guarantee the requirement that the HTLC preimage correspond to a given pubkey we need to introduce a new opcode tentatively called `OP_PUBSECP256K1FROMPRIV`. This new script function would allow us to create a smart contract to ensure that not only a given preimage `x` corresponds to `Hash(x)` but that it also _at the same time_ corresponds to `Pubkey(x)`.
```

Comment:
```suggestion
In order to guarantee the requirement that the HTLC preimage corresponds to a given pubkey we need to introduce a new opcode tentatively called `OP_PUBSECP256K1FROMPRIV`. This new script function would allow us to create a smart contract to ensure that not only a given preimage `x` corresponds to `Hash(x)` but that it also _at the same time_ corresponds to `Pubkey(x)`.
```